python/lsst/pipe/tasks/makeGen3SkyMap.py

```diff
@@ -0,0 +1,131 @@
+            "but does not save it to the data repository",
+        dtype=bool,
+        default=True,
+    )
+
+    def validate(self):
+        if self.name is None:
+            raise ValueError("The name field must be set to the name of the specific "
+                             "skymap to use when writing to the butler")
+
+
+class MakeGen3SkyMapTask(pipeBase.Task):
+    ConfigClass = MakeGen3SkyMapConfig
+    _DefaultName = "makeGen3SkyMap"
+
+    """This is a task to construct and optionally save a SkyMap into a gen3
+    butler repository.
+
+    Parameters
+    ----------
+    config : `MakeGen3SkyMapConfig` or None
+        Instance of a configuration class specifying task options, a default
+        config is created if value is None
+    butler : `str`
+        Path to a gen3 butler
+    collection : `str`
+        The name of the collection which the created skymap will be inserted
+    """
+
+    def __init__(self, *, config=None, butler=None, collection=None, **kwargs):
+        if butler is None:
+            raise ValueError("A path to a butler must be given")
+        if collection is None:
+            raise ValueError("The collection to use with this butler must be provided")
+        super().__init__(config=config, **kwargs)
+        self.butler = Butler(butler, run=collection)
+        self.collection = collection
+
+    def run(self):
+        """Construct and optionally save a SkyMap into a gen3 repository"""
+        skyMap = self.config.skyMap.apply()
+        self.logSkyMapInfo(skyMap)
+        skyMapHash = skyMap.getSha1()
+        if self.config.doWrite:
+            try:
+                existing, = self.butler.registry.query("SELECT skymap FROM skymap WHERE hash=:hash",
+                                                       hash=skyMapHash)
+                raise RuntimeError(
+                    (f"SkyMap with name {existing.name} and hash {skyMapHash} already exist in "
+                     f"the butler collection {self.collection}, SkyMaps must be unique within "
+                     "a collection")
+                )
+            except ValueError:
+                self.log.info(f"Inserting SkyMap {self.config.name} with hash={skyMapHash}")
+                with self.butler.registry.transaction():
+                    skyMap.register(self.config.name, self.butler.registry)
+                    self.butler.registry.registerDatasetType(DatasetType(name=self.config.datasetTypeName,
+                                                                         dimensions=["skymap"],
+                                                                         storageClass="SkyMap"))
+                    self.butler.put(skyMap, self.config.datasetTypeName, {"skymap": self.config.name})
+
+        return pipeBase.Struct(
+            skyMap=skyMap
+        )
+
+    def logSkyMapInfo(self, skyMap):
+        """!Log information about a sky map
+        @param[in] skyMap  sky map (an lsst.skyMap.SkyMap)
+        """
```

Comment:
Should be a numpydoc docstring, and note that the actual class is `lsst.skymap.BaseSkyMap`.

Also, given the duplication with the Gen2 `makeSkyMapTask`, maybe we should move this code into `BaseSkyMap`, as a method that takes a `Log` instance?